From 1e5e8063e67aba3e747e74460e7704a7c39e8598 Mon Sep 17 00:00:00 2001
From: Rui Li <lirui@apache.org>
Date: Wed, 10 May 2017 15:46:31 +0800
Subject: [PATCH 1132/1164] CDH-54016 HIVE-16593: SparkClientFactory.stop may
 prevent JVM from exiting (Rui reviewed by Xuefu)

(cherry picked from commit 57ae3aca05d21628df620f33a9f03966f33c8d7b)

Change-Id: Ie27e5a011f899e76adc6e25e31a59c738e753f05
---
 .../hive/spark/client/SparkClientFactory.java      |   11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientFactory.java b/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientFactory.java
index b23ff2d..cf60b13 100644
--- a/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientFactory.java
+++ b/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientFactory.java
@@ -44,6 +44,7 @@
   static final String CONF_KEY_SECRET = "spark.client.authentication.secret";
 
   private static RpcServer server = null;
+  private static final Object stopLock = new Object();
 
   /**
    * Initializes the SparkClient library. Must be called before creating client instances.
@@ -61,10 +62,14 @@ public static synchronized void initialize(Map<String, String> conf) throws IOEx
   }
 
   /** Stops the SparkClient library. */
-  public static synchronized void stop() {
+  public static void stop() {
     if (server != null) {
-      server.close();
-      server = null;
+      synchronized (stopLock) {
+        if (server != null) {
+          server.close();
+          server = null;
+        }
+      }
     }
   }
 
-- 
1.7.9.5

